ow_regolith/src/RegolithSpawner.cpp

```diff
@@ -0,0 +1,295 @@
+
+  // load SDF model
+  if (!getSdfFromUri(m_model_uri, m_model_sdf)) {
+    ROS_ERROR("Failed to load SDF for regolith model");
+    return false;
+  }
+  // parse as SDF object to query properties about the model
+  auto sdf = parseSdf(m_model_sdf);
+  float model_mass;
+  if (!getModelLinkMass(sdf, model_mass) ||
+      !getModelLinkName(sdf, m_model_linkname)) {
+    ROS_ERROR("Failed to acquire regolith model SDF parameters");
+    return false;
+  }
+
+  // connect to all ROS services
+  ros::ServiceClient gz_get_phys_prop;
+  if (!createRosServiceClient<GetPhysicsProperties>(m_node_handle, 
+        SRV_GET_PHYS_PROPS, gz_get_phys_prop, false) ||
+      !createRosServiceClient<SpawnModel>(m_node_handle, 
+        SRV_SPAWN_MODEL, m_gz_spawn_model, true) ||
+      !createRosServiceClient<DeleteModel>(m_node_handle, 
+        SRV_DELETE_MODEL, m_gz_delete_model, true) ||
+      !createRosServiceClient<ApplyBodyWrench>(m_node_handle, 
+        SRV_APPLY_WRENCH, m_gz_apply_wrench, true) ||
+      !createRosServiceClient<BodyRequest>(m_node_handle, 
+        SRV_CLEAR_WRENCH, m_gz_clear_wrench, true)) {
+    ROS_ERROR("Failed to connect to all required ROS services");
+    return false;
+  }
+
+  // query gazebo for the gravity vector
+  GetPhysicsProperties msg;
+  if (!callRosService(gz_get_phys_prop, msg)) {
+    ROS_ERROR("Failed to query Gazebo for gravity vector");
+    return false;
+  }
+  Vector3 gravity(
+    msg.response.gravity.x, msg.response.gravity.y, msg.response.gravity.z
+  );
+  // set psuedo force to the model's weight multiplied by the weight factor
+  m_psuedo_force_mag 
+    = model_mass * gravity.length() * PSUEDO_FORCE_WEIGHT_FACTOR;
+
+  // subscribe callbacks to ROS topics
+  m_modify_terrain_visual = m_node_handle->subscribe(
+    TOPIC_MODIFY_TERRAIN_VISUAL, 1, &RegolithSpawner::terrainVisualModCb, this);
+  m_dig_linear_result = m_node_handle->subscribe(
+    TOPIC_DIG_LINEAR_RESULT, 1, &RegolithSpawner::armDigLinearResultCb, this);
+  m_dig_circular_result = m_node_handle->subscribe(
+    TOPIC_DIG_CIRCULAR_RESULT, 1, &RegolithSpawner::armDigCircularResultCb, this);
+  m_deliver_result = m_node_handle->subscribe(
+    TOPIC_DELIVER_RESULT, 1, &RegolithSpawner::armDeliverResultCb, this);
+
+  return true;
+}
+
+bool RegolithSpawner::spawnRegolithInScoop(Vector3 pushback_direction)
+{
+  ROS_INFO("Spawning regolith");
+
+  // spawn model
+  static auto spawn_count = 0;
+  stringstream model_name;
+  model_name << "regolith_" << spawn_count++;
+
+  SpawnModel spawn_msg;
+  
+  spawn_msg.request.model_name                  = model_name.str();
+  spawn_msg.request.model_xml                   = m_model_sdf;
+  spawn_msg.request.robot_namespace             = "/regolith";
+  spawn_msg.request.reference_frame             = "lander::l_scoop_tip";
+
+  spawn_msg.request.initial_pose.position.x     = 0.0;
+  spawn_msg.request.initial_pose.position.y     = 0.0;
+  spawn_msg.request.initial_pose.position.z     = -0.05;
```

Comment:
This suggestion is compatible with the new way in which the pushback direction is computed, but I want to hold off on applying until we see further issues with particle retention. I worry it could cause particles to spawn on top of previously spawned particles in instances where the scoop is extremely full.